### PR TITLE
feat: indent on enter after certain EOL tokens

### PR DESCRIPTION
**Pull request by thorimur**
This PR causes indentation when hitting enter after certain tokens that appear at the end of a line.
Specifically, after:
* `by`, `do`, `try`, `finally`, `then`, `else`, `where`, `extends`, `deriving`, `:=`,` =>`, and `<|`

It also inserts two indentations in the case that we hit enter immediately after ` :` on a line that doesn't start with a binder bracket. We exclude binder-bracket-initial lines so as to avoid inserting improper indentation after already-indented hypotheses:

```
theorem foo (...)
    (...) :| <-- hit enter here
    | <-- cursor ends up here
```

If we encounter one of the "postindented EOL tokens" or a trailing ` :` in the first line of a focus block, we add extra indentation as appropriate.

-----

- [ ] depends on: #328 

Ideally, we'd be able to describe indentation using Lean's parser, as an autoformatter implemented by the extension would.

Currently, this adds extra indents (after the first) by appending `\t`. ~~I'm not sure if VS code can be asked to indent twice "natively" in this situation, and I'm not sure if using `\t` would be problematic when it comes to distinctions between space-based indents and tab-based indents.~~ It seems that VS code converts these to spaces as appropriate, and cannot be asked to indent multiple times any other way.

This PR is WIP; please feel free to comment with any opinions or suggestions, especially if I've missed any common tokens! :)

### CODE REVIEW

File: vscode-lean4/language-configuration.json [hunk 1]
```diff
@@ -130,5 +130,39 @@
...
+            "action" : { "indent" : "indent" }
+        },
+        {
+            // Indent thrice after EOL `:` at the start of a focus block
+            "beforeText": "^\\s*(·|\\.)\\s(.*\\s)?:\\s*$",
+            "action" : { "indent" : "indent", "appendText": "\t\t" }
+        },
+        {
+            // Indent twice after EOL `:`
+            "beforeText": "^.*\\s:\\s*$",
```
[the remark this answers is not in the document]
Comment (thorimur) [reply vscode-lean4/language-configuration.json:159]: Hmm—what about only restricting it to lines which don't start with whitespace followed by a parenthesis (or in general, binder bracket)?

This doesn't account for type signatures with really long *hypothesis* type signatures, but those usually require careful manual formatting anyway. The main advantage here is that it reduces the complexity of the regex (there are a lot of different initial tokens to account for!) and is extensible to new tactics with `have`-like syntax (such as `suffices`) and new `def`-like commands (like `irreducible_def`).

File: vscode-lean4/language-configuration.json [hunk 2]
```diff
@@ -130,5 +130,39 @@
...
+        /*
+            Postindented End-Of-Line tokens:
+            The list of English-language EOL tokens which cause indentation on the next line are:
+                by, do, try, finally, then, else, where
+            The list of symbolic EOL tokens which cause indentation on the next line are:
+                :=, =>, <|
+        */
+        {
+            // Indent focus blocks followed by a postindented EOL token twice
+            "beforeText": "^\\s*(·|\\.)\\s(.*\\s)?(by|do|try|finally|then|else|where|\\:=|=>|<\\|)\\s*$",
```
[the remark this answers is not in the document]
Comment (thorimur) [reply vscode-lean4/language-configuration.json:144]: Ah, ok—by the way, do you know if there's a reference somewhere for what regex flavor is used by VS code? I see some hints that it's oniguruma, but I'm not sure. (In this case I can just test it and be assured it works, but I'm curious about the general case. :) )

[the remark this answers is not in the document]
Comment (thorimur) [reply vscode-lean4/language-configuration.json]: Done in 7f89224 and 4828a03 respectively. :)

**Comment by mhuisi** [comment]
(I haven't forgotten this PR and I'll get back to re-reviewing it as soon as I find time for it)

**Comment by mhuisi** [comment]
After mulling on this PR for a while, I think that there are unfortunately too many edge cases where the post-indented EOL token heuristics do the wrong thing. I'm afraid that accidentally adding incorrect indentation on occasion is much more annoying than consistently not adding any additional indentation, so I would prefer to not add these rules.

I've thought a bit about whether there is a better way to add these rules than the one in this PR, but I can't think of any with the limited view of the text that VS Code gives us with this configuration option.

**Comment by thorimur** [comment]
Makes sense! (I wish VS code would let us access the result of parsing somehow...)

Just to be clear, do you consider adding some indentation but not enough indentation to be part of the issue, or is the problem only the potential for "overshooting"? If the latter, I think there are some tokens which are always followed by at *least* one postindentation (such as `try`), for which we could develop rules that never overshoot. (We could also restrict to different circumstances to avoid overshooting, such as "`by` when the line starts with no whitespace and a `def`-like token".) (We'd want to verify any such universal style claims with regexes on mathlib before committing to them, of course.)

But, I undertand if you feel that adding some-but-not-necessarily-enough indentation causes more cognitive overhead than simply not doing so at all. :)

**Comment by mhuisi** [comment]
> Makes sense! (I wish VS code would let us access the result of parsing somehow...)

There's the [textDocument/onTypeFormatting](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_onTypeFormatting) request, but the central problem is that you need to elaborate all dependencies of a declaration in order to be able to parse it, and so the latency for this request can be unacceptably high (think trying to use auto-completion in a portion of the document where the orange bars haven't disappeared yet). Hence, this approach unfortunately isn't really feasible for us, either.

> Just to be clear, do you consider adding some indentation but not enough indentation to be part of the issue, or is the problem only the potential for "overshooting"?

I consider inconsistent behavior in general to be bad and especially bad w.r.t. something that you usually don't need to think about, like input. In this context, I'd say that inconsistently adding too much indentation is worse than inconsistently adding too little indentation, because having to actively revert something that the computer did incorrectly is always extra frustrating (compared to needing to help it along).

Specifically, I don't see a way that we can handle multi-line function signatures correctly in a way that doesn't overshoot or undershoot, and I think that this specific inconsistency will be sufficiently frustrating in practice (in either direction) that it's probably better to err on the side of simple indentation inheritance.

**Comment by mhuisi** [comment]
Closing this for now due to reasons mentioned above.

**Pull request closed without merging**